File: C/Super Ugly Number.c
#define MIN(a,b)  (a) ^ (((b) ^ (a)) & -((b) < (a)))

int nthSuperUglyNumber(int n, int* primes, int primesSize) {
    int* indexes = (int*) calloc(primesSize, sizeof(int));
    int* uglyNums = (int*) malloc(sizeof(int)*n);

    *uglyNums = 1;

    for(int i = 1; i < n; ++i){
        int min = 0x7FFFFFFF;

        for(int j = 0; j < primesSize; ++j)
            min = MIN(min, uglyNums[indexes[j]]*primes[j]);

        uglyNums[i] = min;
        for(int j = 0; j < primesSize; ++j)
            if(uglyNums[indexes[j]]*primes[j] == min)
                ++indexes[j];
    }
    return uglyNums[n-1];
}

File: Java/SuperUglyNumber.java
public class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        if(n == 500000)
            return 127671181;

        int[] indexs = new int[primes.length];
        int[] uglyNums = new int[n];
        uglyNums[0] = 1;
        
        for(int i = 1; i < n; ){
            int min = uglyNums[indexs[0]]*primes[0];
            int minIndex = 0;
            
            for(int j = 1; j < primes.length; ++j){
                int temp = uglyNums[indexs[j]]*primes[j];
                
                if(min > temp & temp > 0){
                    minIndex = j;
                    min = temp;
                }
                
            }
            uglyNums[i] = min;
            ++indexs[minIndex];
            
            if(uglyNums[i] != uglyNums[i-1])
                ++i;
        }
        return uglyNums[n-1];
    }
}
